include/sunlinsol/sunlinsol_ginkgobatch.hpp

```diff
@@ -0,0 +1,505 @@
+                        precon_factory, max_iters, num_batches, sunctx)
+  {}
+
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  BatchLinearSolver(std::shared_ptr<const gko::Executor> gko_exec,
+                    gko::batch::stop::tolerance_type tolerance_type,
+                    std::shared_ptr<gko::batch::BatchLinOpFactory> precon_factory,
+                    gko::size_type num_batches, SUNContext sunctx)
+    : BatchLinearSolver(std::move(gko_exec), tolerance_type, precon_factory,
+                        default_max_iters_, num_batches, sunctx)
+  {}
+
+  /// Get the ``gko::Executor`` associated with the Ginkgo matrix
+  std::shared_ptr<const gko::Executor> GkoExec() const { return gko_exec_; }
+
+  // Override the ConvertibleTo methods
+
+  /// Implicit conversion to a :c:type:`SUNLinearSolver`
+  operator SUNLinearSolver() override { return object_.get(); }
+
+  /// Implicit conversion to a :c:type:`SUNLinearSolver`
+  operator SUNLinearSolver() const override { return object_.get(); }
+
+  /// Explicit conversion to a :c:type:`SUNLinearSolver`
+  SUNLinearSolver Convert() override { return object_.get(); }
+
+  /// Explicit conversion to a :c:type:`SUNLinearSolver`
+  SUNLinearSolver Convert() const override { return object_.get(); }
+
+  /// Get the underlying Ginkgo solver
+  /// \note This will be `nullptr` until the linear solver setup phase.
+  GkoBatchSolverType* GkoSolver() { return solver_.get(); }
+
+  /// Average number of iterations across the batches during the last solve.
+  sunrealtype AvgNumIters() const { return avg_iter_count_; }
+
+  /// Standard deviation of the number of iterations across the batches during the last solve.
+  sunrealtype StddevNumIters() const { return stddev_iter_count_; }
+
+  /// Running sum of the average number of iterations in this solvers lifetime.
+  sunrealtype SumAvgNumIters() const { return sum_of_avg_iters_; }
+
+  void SetScalingMode(int scaling_mode = LAGGED_SCALING)
+  {
+    scaling_mode_ = scaling_mode;
+  }
+
+  /// Sets the left and right scaling vectors to be used.
+  void SetScalingVectors(N_Vector s1, N_Vector s2)
+  {
+    SUNFunctionBegin(sunCtx());
+
+    if (s1 && s2)
+    {
+      s1_ = s1;
+      col_scale_vec_ =
+        std::move(impl::WrapBatchScalingArray(GkoExec(), num_batches_, s1_));
+
+      if (!s2inv_.Convert())
+      {
+        s2inv_ = sundials::experimental::NVectorView(N_VClone(s2));
+      }
+
+      // SUNLinearSolver API wants s2inv_
+      N_VInv(s2, s2inv_);
+      SUNCheckLastErrVoid();
+
+      row_scale_vec_ =
+        std::move(impl::WrapBatchScalingArray(GkoExec(), num_batches_, s2inv_));
+
+      scaling_initialized_ = true;
+    }
+    else
+    {
+      scaling_mode_        = NO_SCALING;
+      scaling_initialized_ = false;
+    }
+  }
+
+  int Setup(BatchMatrix<GkoBatchMatType>* A)
+  {
+    if (num_batches_ != A->NumBatches()) { return SUN_ERR_ARG_OUTOFRANGE; }
+
+    matrix_   = A;
```

Comment:
Rather than saving `A` here, I think it would be a little cleaner to just set the `do_setup_` flag and pass `A` through to `Solve`.